docs/proposals/secret-configmap-file-mode.md

```diff
@@ -0,0 +1,159 @@
+   if we don't want to change kubernetes core for some reason, although there
+   seem to be valid use cases. But if the user want's to use the "postStart" for
+   something else, then it is more disturbing to do both things (have a script
+   in the docker image that deals with this, but is not probably concern of the
+   project so it's not nice, or specify several commands by using "sh").
+
+# Proposed implementation
+
+The proposed implementation goes with the first alternative: adding a `mode`
+to the API.
+
+There will be a `defaultMode`, type `int`, in `type SecretVolumeSource` and
+`type ConfigMapVolumeSource` and a `mode`, type `int` too, in `type KeyToPath`.
+The mask provided in any of these fields will be ANDed with 0777 to disallow
+setting sticky and setuid bits. It's not clear that use case is needed nor
+really understood.
+
+Directories within the volume will be created as before and are not affected by
+this setting.
+
+In other words, the fields will look like this (will be improved during the
+implementation):
+```
+type SecretVolumeSource struct {
+        // Name of the secret in the pod's namespace to use.
+        SecretName string `json:"secretName,omitempty"`
+        // If unspecified, each key-value pair in the Data field of the referenced
+        // Secret will be projected into the volume as a file whose name is the
+        // key and content is the value. If specified, the listed keys will be
+        // projected into the specified paths, and unlisted keys will not be
+        // present. If a key is specified which is not present in the Secret,
+        // the volume setup will error. Paths must be relative and may not contain
+        // the '..' path or start with '..'.
+        Items       []KeyToPath `json:"items,omitempty"`
+        // Mode bits to use on created files by default. The used mode bits will
+        // be the provided AND 0777.
+        // Directories within the path are not affected by this setting
+        DefaultMode int         `json:"defaultMode,omitempty"`
+}
+
+type ConfigMapVolumeSource struct {
+        LocalObjectReference `json:",inline"`
+        // If unspecified, each key-value pair in the Data field of the referenced
+        // ConfigMap will be projected into the volume as a file whose name is the
+        // key and content is the value. If specified, the listed keys will be
+        // projected into the specified paths, and unlisted keys will not be
+        // present. If a key is specified which is not present in the ConfigMap,
+        // the volume setup will error. Paths must be relative and may not contain
+        // the '..' path or start with '..'.
+        Items       []KeyToPath `json:"items,omitempty"`
+        // Mode bits to use on created files by default. The used mode bits will
+        // be the provided AND 0777.
+        // Directories within the path are not affected by this setting
+        DefaultMode int         `json:"defaultMode,omitempty"`
+}
```

Comment:
On Mon, Jul 11, 2016 at 10:28:57PM -0700, Tim Hockin wrote:
> > +
> > +type ConfigMapVolumeSource struct {
> > +        LocalObjectReference `json:",inline"`
> > +        // If unspecified, each key-value pair in the Data field of the referenced
> > +        // ConfigMap will be projected into the volume as a file whose name is the
> > +        // key and content is the value. If specified, the listed keys will be
> > +        // projected into the specified paths, and unlisted keys will not be
> > +        // present. If a key is specified which is not present in the ConfigMap,
> > +        // the volume setup will error. Paths must be relative and may not contain
> > +        // the '..' path or start with '..'.
> > +        Items       []KeyToPath `json:"items,omitempty"`
> > +        // Mode bits to use on created files by default. The used mode bits will
> > +        // be the provided AND 0777.
> > +        // Directories within the path are not affected by this setting
> > +        DefaultMode int         `json:"defaultMode,omitempty"`
> > +}
> 
> and DownwardAPIVolumeFile

Sure! I'll check if it is needed in some other place too, just in case :-)